altair/utils/core.py

```diff
@@ -118,115 +144,116 @@ def _parse_shorthand(shorthand):
-    ----------
-    shorthand: str
-        Shorthand string of the form "agg(col):typ"
-    data : pd.DataFrame (optional)
-        Dataframe from which to infer types
+    >>> parse_shorthand('sum(bar)', data) == {'aggregate': 'sum', 'field': 'bar', 'type': 'quantitative'}
+    True
 
-    Returns
-    -------
-    D : dict
-        Dictionary which always contains a 'field' key, and additionally
-        contains an 'aggregate' and 'type' key depending on the input.
+    >>> parse_shorthand('count()', data) == {'aggregate': 'count', 'type': 'quantitative'}
+    True
+    """
+    if not shorthand:
+        return {}
 
-    Examples
-    --------
-    >>> data = pd.DataFrame({'foo': ['A', 'B', 'A', 'B'],
-    ...                      'bar': [1, 2, 3, 4]})
+    valid_typecodes = list(TYPECODE_MAP) + list(INV_TYPECODE_MAP)
+
+    units = dict(field='(?P<field>.*)',
+                 type='(?P<type>{0})'.format('|'.join(valid_typecodes)),
+                 count='(?P<aggregate>count)',
+                 aggregate='(?P<aggregate>{0})'.format('|'.join(AGGREGATES)),
+                 timeUnit='(?P<timeUnit>{0})'.format('|'.join(TIMEUNITS)))
+
+    patterns = []
 
-    >>> parse_shorthand('name')
-    {'field': 'name'}
+    if parse_aggregates:
+        patterns.extend([r'{count}\(\)',
+                         r'{aggregate}\({field}\)'])
+    if parse_timeunits:
+        patterns.extend([r'{timeUnit}\({field}\)'])
 
-    >>> parse_shorthand('average(col)')  # doctest: +SKIP
-    {'aggregate': 'average', 'field': 'col'}
+    patterns.extend([r'{field}'])
 
-    >>> parse_shorthand('foo:O')  # doctest: +SKIP
-    {'field': 'foo', 'type': 'ordinal'}
+    if parse_types:
+        patterns = list(itertools.chain(*((p + ':{type}', p) for p in patterns)))
 
-    >>> parse_shorthand('min(foo):Q')  # doctest: +SKIP
-    {'aggregate': 'min', 'field': 'foo', 'type': 'quantitative'}
+    regexps = (re.compile('\A' + p.format(**units) + '\Z', re.DOTALL)
+               for p in patterns)
 
-    >>> parse_shorthand('foo', data)  # doctest: +SKIP
-    {'field': 'foo', 'type': 'nominal'}
+    # find matches depending on valid fields passed
+    if isinstance(shorthand, dict):
+        attrs = shorthand
+    else:
+        attrs = next(exp.match(shorthand).groupdict() for exp in regexps
+                     if exp.match(shorthand))
 
-    >>> parse_shorthand('bar', data)  # doctest: +SKIP
-    {'field': 'bar', 'type': 'quantitative'}
+    # Handle short form of the type expression
+    if 'type' in attrs:
+        attrs['type'] = INV_TYPECODE_MAP.get(attrs['type'], attrs['type'])
 
-    >>> parse_shorthand('bar:O', data)  # doctest: +SKIP
-    {'field': 'bar', 'type': 'ordinal'}
+    # counts are quantitative by default
+    if attrs == {'aggregate': 'count'}:
+        attrs['type'] = 'quantitative'
 
-    >>> parse_shorthand('sum(bar)', data)  # doctest: +SKIP
-    {'aggregate': 'sum', 'field': 'bar', 'type': 'quantitative'}
+    # times are temporal by default
```

Comment:
I noticed that right now, if the user provides a `timeUnit` and a non temporal type, we accept that (something like `month(column):N`. In cases like that I wonder if we should override the type with temporal? Or maybe raise a `TypeError`?

Reply:
No, I don't think we should prevent users from using other types. This is something I only appreciated recently: the type specification of a timeUnit refers not to the input data, but to the way the output binned data is treated. So often ordinal or nominal is the right choice (see, e.g. the chart at the top of this PR).

Reply:
How would the above plot change if the type was set to T or N? I guess if you are encoding a datetime using color, that would be an example of possibly wanting to choose O/N instead of T?

Reply:
Hmm, while I was swimming I tried to work through the different options for how these things might work. My initial guess was that for position like channels T behaves as O. However the behavior you are showing here is *very* different. Here is looks like it is treating the datetimes as a range rather than a single point. I almost wonder if that would be considered a bug in VegaLite.

@domoritz - any thoughts.

However, I see your point that we don't want to force datetime aggregations to to have a temporal type.

Reply:
> Here is looks like it is treating the datetimes as a range rather than a single point. I almost wonder if that would be considered a bug in VegaLite.

If you use point mark, it will be a single point.  However, rect mark expects a range input (x,x2,y,y2).  Since x2,y2 are not specified, x2 and y2 are default to minimum values.

Reply:
** However, rect mark expects a range input (x,x2,y,y2) or a discrete value (ordinal/nominal).

Reply:
`temporal` behaves like `quantitative` but some of the defaults around formatting are different. `timeUnit` is a discretization like `bin`. 

@jakevdp's comment about types is exactly hitting this point

> I don't think we should prevent users from using other types. This is something I only appreciated recently: the type specification of a timeUnit refers not to the input data, but to the way the output binned data is treated.

For binning, here are two examples where we use Q and O and you can see how the output changes: https://vega.github.io/vega-lite/docs/bin.html#histogram and https://vega.github.io/vega-lite/docs/bin.html#histogram-ordinal. Note that when the type is O, the bars can also be sorted by some other variable. 

Time Unit has similar behavior: https://vega.github.io/vega-lite/docs/timeunit.html. There are certainly a few smart defaults you can make when you know the raw type, the mark type, and the channel but I would not force the user to use a particular type. We try to add warnings in Vega-Lite for cases where you have accidental overlap for example.

Reply:
Thanks I hadn't appreciated some of these subtle points about the rect mark and temporal datatype. Sounds like this PR is solid on that front end, and just needs a few tests.